### PR TITLE
Performance improvement for C++ rrtmgpxx

### PR DESCRIPTION
**Pull request by mrnorman**
This PR fixes performance problems in the YAKL C++ port of RRTMGP. The main improvements come from inlining interpolate2D and interpolate3D. My benchmarks on the CPU on my laptop indicate the C++ LW is now 17% more expensive than the OpenACC Fortran code and the C++ SW is now 34% more expensive than the Fortran OpenACC code. This is a marked improvement over previous runtimes and brings the YAKL RRTMGP port back to the performance it used to have compared to Fortran OpenACC code on the CPU.

**This PR requires the latest YAKL main branch**.

Change cmake to process targets. Replacing rrtmgpxx timers with YAKL timers. Inlining interpolate3D and interpolate2D to improve performance and remove array slicing overheads. Change anyLT and anyGT to more fortran-like syntax with the new yakl componentwise operator overloads and intrinsic, any. Changing extract_subset_dim2_4d to accept 4-D arrays instead of 3-D (bug fix for previous implementation). Commenting out expand() in mo_optical_props.h because it contains code with bugs I don't know how to fix. Changing C++ testing framework to compile Fortran code with cmake with and without openacc code. The original fortran build system was a struggle. Changing testing cmake to link netcdf fortran. Getting rid of YAKL_DEVICE_LAMBDA. It isn't needed any more.

### CODE REVIEW

**Comment by mrnorman** [comment]
Since this requires the latest YAKL, I'm happy to help integrate into E3SM-MMF. It *might* require changes in the samxx code.

File: cpp/rte/kernels/mo_fluxes_broadband_kernels.cpp [hunk 1]
```diff
@@ -14,18 +17,22 @@ void sum_broadband(int ncol, int nlev, int ngpt, real3d const &spectral_flux, re
...
+  parallel_for( SimpleBounds<2>(nlev,ncol) , YAKL_LAMBDA (int ilev, int icol) {
     real diff = spectral_flux_dn(icol, ilev, 1) - spectral_flux_up(icol, ilev, 1);
     broadband_flux_net(icol, ilev) = diff;
   });
 
   // do igpt = 2, ngpt
   //   do ilev = 1, nlev
   //     do icol = 1, ncol
-  parallel_for( Bounds<3>({2,ngpt},nlev,ncol) , YAKL_DEVICE_LAMBDA (int igpt, int ilev, int icol) {
+  parallel_for( Bounds<3>({2,ngpt},nlev,ncol) , YAKL_LAMBDA (int igpt, int ilev, int icol) {
```
Comment (brhillman) [inline cpp/rte/kernels/mo_fluxes_broadband_kernels.cpp:35]: Do we no longer need `YAKL_DEVICE_LAMBDA` for kernels with atomic reductions?

File: cpp/rte/mo_optical_props.h [hunk 2]
```diff
@@ -216,21 +232,21 @@ class OpticalProps {
...
-    parallel_for( Bounds<1>(1) , YAKL_LAMBDA (int dummy) {
-      for (int iband = 1 ; iband <= nband ; iband++) {
-        for (int i=this_band2gpt(1,iband) ; i <= this_band2gpt(2,iband) ; i++) {
-          ret(i) = arr_in(iband);
-        }
-      }
-    });
-    return ret;
-  }
+  // real1d expand(real1d const &arr_in) const {
```
Comment (brhillman) [inline cpp/rte/mo_optical_props.h:235]: Is this intended to be commented out? Should it just be removed?

File: cpp/rte/mo_optical_props.h [hunk 3]
```diff
@@ -343,14 +372,20 @@ class OpticalProps2str : public OpticalPropsArry {
       stoprun("validate: arrays not sized consistently");
     }
     #ifdef RRTMGP_EXPENSIVE_CHECKS
-      if (anyLT(this->tau, 0._wp)                          ) { stoprun("validate: tau values out of range"); }
-      if (anyLT(this->ssa, 0._wp) || anyGT(this->ssa,1._wp)) { stoprun("validate: ssa values out of range"); }
-      if (anyLT(this->g  ,-1._wp) || anyGT(this->g  ,1._wp)) { stoprun("validate: g   values out of range"); }
+      if (any(this->tau <  0)                      ) { stoprun("validate: tau values out of range"); }
```
Comment (brhillman) [inline cpp/rte/mo_optical_props.h:375]: These comparisons no longer have explicit precision, is this okay?

**Review by brhillman — APPROVED** [review]
This looks good. Having to inline the interpolate2d and interpolate3d calls is unfortunate, but not sure there is a way around that.

**Pull request merged**